gcc/rust/rust-diagnostics.cc

```diff
@@ -199,13 +199,16 @@ class rust_error_code_rule : public diagnostic_metadata::rule
 
   char *make_description () const final override
   {
-    return xstrdup (m_code.m_str);
+    std::string errstring (ErrorCodeString.at (m_code));
+    char *desc = new char[errstring.size () + 1];
+    std::strcpy (desc, errstring.c_str ());
+    return desc;
```

Comment:
```suggestion
    return ErrorCodeString.at (m_code);
```

this should work fine. It might require you to change the definition of the `ErrorCodeString` table to...

Reply:
But then we have to convert the returned string of the stringizing operator to `char *`.Which I am not sure how to do that? Do you have any idea?

Reply:
I've tried, and there's no need to do anything like that. You need to change the type contained in the map as I suggested [here](https://github.com/Rust-GCC/gccrs/pull/2468#discussion_r1276341503) and then it all works. I've tried it locally :)

Reply:
We have talked about it, the returned object of `stringizing operator` is of string type, 

<details><summary>When I try to compile with char * instead of const char* it gives warning and aborts: </summary>
<p>

```bash
../../gccrs/gcc/rust/rust-diagnostics.h:1090:1: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
../../gccrs/gcc/rust/rust-diagnostics.h:1090:1: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
../../gccrs/gcc/rust/rust-diagnostics.h:1090:1: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
In file included from ../../gccrs/gcc/rust/ast/rust-ast.h:27,
                 from ../../gccrs/gcc/rust/ast/rust-ast-full.h:22,
                 from ../../gccrs/gcc/rust/ast/rust-ast-formatting.cc:20:
../../gccrs/gcc/rust/rust-diagnostics.h:584:19: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
  584 | TABLE_TO_MAP(E0001),
      |                   ^
../../gccrs/gcc/rust/rust-diagnostics.h:579:17: note: in definition of macro ‘STR’

```


</p>
</details> 

<details><summary>If I try to use string, then the return type of function doesn't match & it gives this error:
</summary>
<p>

```bash
../../gccrs/gcc/rust/rust-diagnostics.cc: In member function ‘virtual char* rust_error_code_rule::make_description() const’:
../../gccrs/gcc/rust/rust-diagnostics.cc:202:49: error: invalid conversion from ‘const char*’ to ‘char*’ [-fpermissive]
  202 |     return error_code_strings.at (m_code).c_str ();
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
      |                                                 |
      |                                                 const char*
../../gccrs/gcc/rust/rust-diagnostics.cc: In member function ‘virtual char* rust_error_code_rule::make_url() const’:
../../gccrs/gcc/rust/rust-diagnostics.cc:207:19: warning: passing objects of non-trivially-copyable type ‘std::unordered_map<ErrorCode, std::__cxx11::basic_string<char> >::mapped_type’ {aka ‘class std::__cxx11::basic_string<char>’} through ‘...’ is conditionally supported [-Wconditionally-supported]
  207 |     return concat ("https://doc.rust-lang.org/error-index.html#",
      |            ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  208 |                    error_code_strings.at (m_code), NULL);
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```





</p>
</details> 



We need to send a copy of char * to each of the function, because we are calling free from both function (make_description & make_url), can be seen here: https://github.com/Rust-GCC/gccrs/blob/master/gcc/diagnostic.cc#L1376-L1379,

So, I have updated the PR with this


```c++
  char *make_description () const final override
  {
    return xstrdup (error_code_strings.at (m_code));
  }
```